ruby-install-0.9.3/doc/man/ruby-install.1.md
# ruby-install 1 "Aug 2018" ruby-install "User Manuals"

## SYNOPSIS

`ruby-install` [OPTIONS] [[RUBY \| VERSION \| RUBY-VERSION] [-- CONFIGURE_OPTS...]]

## DESCRIPTION

Installs Ruby, JRuby, TruffleRuby, or mruby.

https://github.com/postmodern/ruby-install#readme

## ARGUMENTS

*RUBY*
	Install Ruby by name. Must be either `ruby`, `jruby`, `truffleruby`, or
  `mruby`.

*VERSION*
	Optionally select the version of selected Ruby.

*CONFIGURE_OPTS*
	Additional optional configure arguments.

## OPTIONS

`-r`, `--rubies-dir` *DIR*
	Specifies the alternate directory where other Ruby directories are
	installed, such as *~/.rvm/rubies* or *~/.rbenv/versions*.
	Defaults to */opt/rubies* for root and *~/.rubies* for normal users.

`-i`, `--install-dir` *DIR*
	Specifies the directory where Ruby will be installed.
	Defaults to */opt/rubies/$ruby-$version* for root and
	*~/.rubies/$ruby-$version* for normal users.

`--prefix` *DIR*
	Alias for `-i DIR`.

`--system`
	Alias for `-i /usr/local`.

`-s`, `--src-dir` *DIR*
	Specifies the directory for downloading and unpacking Ruby source.

`-c`, `--cleanup`
	Remove the downloaded Ruby archive and unpacked source-code after
	installation.

`-j`, `--jobs` *NUM*, `-j`*NUM*, `--jobs=`*NUM*
	Specifies the number of *make* jobs to run in parallel when compiling
	Ruby. If the -j option is provided without an argument, *make* will
	allow an unlimited number of simultaneous jobs.

`-p`, `--patch` *FILE*
	Specifies any additional patches to apply.

`-M`, `--mirror` *URL*
	Specifies an alternate mirror to download the Ruby archive from.

`-u`, `--url` *URL*
	Alternate URL to download the Ruby archive from.

`-m`, `--md5` *MD5*
	Specifies the MD5 checksum for the Ruby archive.

`--sha1` *SHA1*
	Specifies the SHA1 checksum for the Ruby archive.

`--sha256` *SHA256*
	Specifies the SHA256 checksum for the Ruby archive.

`--sha512` *SHA512*
	Specifies the SHA512 checksum for the Ruby archive.

`--package-manager [apt|dnf|yum|pacman|zypper|brew|pkg|port]`
  Use an alternative package manager.

`--no-download`
	Use the previously downloaded Ruby archive.

`--no-verify`
	Do not verify the downloaded Ruby archive.

`--no-extract`
	Do not extract the downloaded Ruby archive. Implies `--no-download`
	and `--no-verify`.

`--no-install-deps`
	Do not install build dependencies before installing Ruby.

`--no-reinstall`
	Skip installation if another Ruby is detected in same location.

`-U`, `--update`
	Downloads the latest ruby versions and checksums from the ruby-versions
	repository (https://github.com/postmodern/ruby-versions#readme).

`-D`, `--debug`
    Enable debugging messages.

`-V`, `--version`
	Prints the current ruby-install version.

`-h`, `--help`
	Prints a synopsis of ruby-install usage.

## EXAMPLES

List supported Rubies and their major versions:

    $ ruby-install

List the latest versions:

    $ ruby-install --latest

Install the current stable version of Ruby:

    $ ruby-install ruby

Install the latest version of Ruby:

    $ ruby-install --latest ruby

Install a latest version of Ruby:

    $ ruby-install ruby 3.1

Install a specific version of Ruby:

    $ ruby-install ruby 3.1.2

Install a Ruby into a specific directory:

    $ ruby-install --install-dir /path/to/dir ruby

Install a Ruby into a specific `rubies` directory:

    $ ruby-install --rubies-dir /path/to/rubies/ ruby

Install a Ruby into `/usr/local`:

    $ ruby-install --system ruby 3.1.2

Install a Ruby from an official site with directly download:

    $ ruby-install -M https://ftp.ruby-lang.org/pub/ruby ruby 3.1.2

Install a Ruby from a mirror:

    $ ruby-install -M http://www.mirrorservice.org/sites/ftp.ruby-lang.org/pub/ruby ruby 2.4.0

Install a Ruby with a specific patch:

    $ ruby-install -p https://raw.github.com/gist/4136373/falcon-gc.diff ruby 1.9.3-p551

Install a Ruby with specific configuration:

    $ ruby-install ruby 3.1.2 -- --enable-shared --enable-dtrace CFLAGS="-O3"

Using ruby-install with [RVM]:

    $ ruby-install --rubies-dir ~/.rvm/rubies ruby 3.1.2

Using ruby-install with [rbenv]:

    $ ruby-install -i ~/.rbenv/versions/2.4.0 ruby 3.1.2

Uninstall a Ruby version:

    $ rm -rf ~/.rubies/ruby-3.1.2

## FILES

*/usr/local/src*
	Default root user source directory.

*~/src*
	Default non-root user source directory.

*/opt/rubies/$ruby-$version*
	Default root user installation directory.

*~/.rubies/$ruby-$version*
	Default non-root user installation directory.

## ENVIRONMENT

*RUBY_INSTALL_SRC_DIR*
  Overrides the default source directory.

*RUBY_INSTALL_RUBIES_DIR*
  Overrides the default installation directory.

*RUBY_INSTALL_PKG_MANAGER*
  Explicitly sets the package manager to use.

## AUTHOR

Postmodern <postmodern.mod3@gmail.com>

## SEE ALSO

ruby(1), gem(1), chruby(1), chruby-exec(1)


_pages/news.md
---
layout: page
title: news
permalink: /news/
description: Activities and news from the Electronic Visualization Laboratory
nav: true
nav_order: 2
---

{% include news.liquid %}


ruby-install-0.9.3/share/ruby-install/ruby-install.sh
#!/usr/bin/env bash

shopt -s extglob

ruby_install_version="0.9.3"
ruby_install_dir="${BASH_SOURCE[0]%/*}"
ruby_install_cache_dir="${XDG_CACHE_HOME:-$HOME/.cache}/ruby-install"

rubies=(ruby jruby truffleruby truffleruby-graalvm mruby)
patches=()
configure_opts=()
make_opts=()

system_dir="/usr/local"

if (( UID == 0 )); then
	src_dir="${RUBY_INSTALL_SRC_DIR:-$system_dir/src}"
	rubies_dir="${RUBY_INSTALL_RUBIES_DIR:-/opt/rubies}"
else
	src_dir="${RUBY_INSTALL_SRC_DIR:-$HOME/src}"
	rubies_dir="${RUBY_INSTALL_RUBIES_DIR:-$HOME/.rubies}"
fi

source "$ruby_install_dir/logging.sh"
source "$ruby_install_dir/system.sh"
source "$ruby_install_dir/package_manager.sh"
source "$ruby_install_dir/util.sh"
source "$ruby_install_dir/ruby-versions.sh"

#
# Prints usage information for ruby-install.
#
function usage()
{
	cat <<USAGE
usage: ruby-install [OPTIONS] [[RUBY|VERSION|RUBY-VERSION] [-- CONFIGURE_OPTS ...]]

Options:

	-r, --rubies-dir DIR	Directory that contains other installed Rubies
	-i, --install-dir DIR	Directory to install Ruby into
	    --prefix DIR        Alias for -i DIR
	    --system		Alias for -i $system_dir
	-s, --src-dir DIR	Directory to download source-code into
	-c, --cleanup		Remove archive and unpacked source-code after installation
	-j, --jobs NUM, -jNUM, --jobs=NUM
				Number of jobs to run in parallel when compiling
	-p, --patch FILE	Patch to apply to the Ruby source-code
	-M, --mirror URL	Alternate mirror to download the Ruby archive from
	-u, --url URL		Alternate URL to download the Ruby archive from
	-m, --md5 MD5		MD5 checksum of the Ruby archive
	    --sha1 SHA1		SHA1 checksum of the Ruby archive
	    --sha256 SHA256	SHA256 checksum of the Ruby archive
	    --sha512 SHA512	SHA512 checksum of the Ruby archive
	--package-manager [apt|dnf|yum|pacman|zypper|brew|pkg|port]
				Use an alternative package manager
	--no-download		Use the previously downloaded Ruby archive
	--no-verify		Do not verify the downloaded Ruby archive
	--no-extract		Do not re-extract the downloaded Ruby archive
	--no-install-deps	Do not install build dependencies before installing Ruby
	--no-reinstall  	Skip installation if another Ruby is detected in same location
	-U, --update		Downloads the latest ruby versions and checksums
	-L, --latest		Deprecated: please use -U,--update instead
	-D, --debug		Enable debug messages
	-V, --version		Prints the version
	-h, --help		Prints this message

Examples:

	$ ruby-install ruby
	$ ruby-install ruby 2.3
	$ ruby-install ruby 2.3.0
	$ ruby-install ruby -- --with-openssl-dir=...
	$ ruby-install -M https://ftp.ruby-lang.org/pub/ruby ruby
	$ ruby-install -M http://www.mirrorservice.org/sites/ftp.ruby-lang.org/pub/ruby ruby
	$ ruby-install -p https://raw.github.com/gist/4136373/falcon-gc.diff ruby 1.9.3

USAGE
}

#
# Parses a "ruby-version" string.
#
function parse_ruby()
{
	local string="$1"

	case "$string" in
		*-[0-9]*)
			ruby="${string%-[0-9]*}"
			ruby_version="${string#$ruby-}"
			;;
		[0-9]*)
			ruby="ruby"
			ruby_version="$string"
			;;
		*)
			ruby="$string"
			;;
	esac

	if [[ ! "${rubies[*]}" == *"$ruby"* ]]; then
		echo "ruby-install: unknown ruby: $ruby" >&2
		return 1
	fi
}

#
# Parses command-line options for ruby-install.
#
function parse_options()
{
	local argv=()

	while [[ $# -gt 0 ]]; do
		case "$1" in
			-r|--rubies-dir)
				rubies_dir="$2"
				shift 2
				;;
			-i|--install-dir|--prefix)
				install_dir="$2"
				shift 2
				;;
			--system)
				install_dir="$system_dir"
				shift 1
				;;
			-s|--src-dir)
				src_dir="$2"
				shift 2
				;;
			-c|--cleanup)
				cleanup=1
				shift
				;;
			-j|--jobs)
				make_opts+=("$1" "$2")
				shift 2
				;;
			-j+([0-9])|--jobs=+([0-9]))
				make_opts+=("$1")
				shift
				;;
			-p|--patch)
				patches+=("$2")
				shift 2
				;;
			-M|--mirror)
				ruby_mirror="$2"
				shift 2
				;;
			-u|--url)
				ruby_url="$2"
				ruby_archive="${ruby_url##*/}"
				shift 2
				;;
			-m|--md5)
				ruby_md5="$2"
				shift 2
				;;
			--sha1)
				ruby_sha1="$2"
				shift 2
				;;
			--sha256)
				ruby_sha256="$2"
				shift 2
				;;
			--sha512)
				ruby_sha512="$2"
				shift 2
				;;
			--package-manager)
				set_package_manager "$2"
				shift 2
				;;
			--no-download)
				no_download=1
				shift
				;;
			--no-verify)
				no_verify=1
				shift
				;;
			--no-extract)
				no_download=1
				no_verify=1
				no_extract=1
				shift
				;;
			--no-install-deps)
				no_install_deps=1
				shift
				;;
			--no-reinstall)
				no_reinstall=1
				shift
				;;
			-L|--latest)
				warn "DEPRECATION: -L,--latest is deprecated. Please use -U,--update instead."
				force_update=1
				shift
				;;
			-U|--update)
				force_update=1
				shift
				;;
			-D|--debug)
				enable_debug=1
				shift
				;;
			-V|--version)
				echo "ruby-install: $ruby_install_version"
				exit
				;;
			-h|--help)
				usage
				exit
				;;
			--)
				shift
				configure_opts=("$@")
				break
				;;
			-*)
				echo "ruby-install: unrecognized option $1" >&2
				return 1
				;;
			*)
				argv+=("$1")
				shift
				;;
		esac
	done

	case ${#argv[*]} in
		2)	parse_ruby "${argv[0]}-${argv[1]}" || return $? ;;
		1)	parse_ruby "${argv[0]}" || return $? ;;
		0)	return 0 ;;
		*)
			echo "ruby-install: too many arguments: ${argv[*]}" >&2
			usage 1>&2
			return 1
			;;
	esac
}

#
# Prints Rubies supported by ruby-install.
#
function list_rubies()
{
	local ruby

	for ruby in "${rubies[@]}"; do
		if [[ $force_update -eq 1 ]] ||
		   are_ruby_versions_missing "$ruby"; then
			log "Downloading latest $ruby versions ..."
			download_ruby_versions "$ruby"
		fi
	done

	echo "Stable ruby versions:"
	for ruby in "${rubies[@]}"; do
		echo "  $ruby:"
		stable_ruby_versions "$ruby" | sed -e 's/^/    /' || return $?
	done
}

#
# Initializes variables.
#
function init()
{
	local fully_qualified_version="$(lookup_ruby_version "$ruby" "$ruby_version")"

	if [[ -n "$fully_qualified_version" ]]; then
		ruby_version="$fully_qualified_version"
	else
		warn "Unknown $ruby version $ruby_version. Proceeding anyways ..."
	fi

	source "$ruby_install_dir/functions.sh"       || return $?
	source "$ruby_install_dir/$ruby/functions.sh" || return $?

	ruby_cache_dir="$ruby_install_cache_dir/$ruby"
	install_dir="${install_dir:-$rubies_dir/$ruby-$ruby_version}"
	ruby_build_dir="$src_dir/$ruby_dir_name"

	ruby_md5="${ruby_md5:-$(ruby_checksum_for "$ruby" md5 "$ruby_archive")}"
	ruby_sha1="${ruby_sha1:-$(ruby_checksum_for "$ruby" sha1 "$ruby_archive")}"
	ruby_sha256="${ruby_sha256:-$(ruby_checksum_for "$ruby" sha256 "$ruby_archive")}"
	ruby_sha512="${ruby_sha512:-$(ruby_checksum_for "$ruby" sha512 "$ruby_archive")}"
}
